README.md
# ISO-Compress
Create your own iso image, and some packages like (Shutdown computer, convert images)
program with class, Python,
help me to optimize it..
Environment: Windows
requirements : Python 3.x,

rynnika@gmail.com



history
Monday 22 February 2016 23:32:41
3600
Restart



convert.py
import tkinter,os
from tkinter.filedialog import askopenfile,askdirectory
from tkinter.messagebox import showinfo
import tkinter.ttk

class Convert():
    def __init__(self):        
        self.bg(self.convert,5,20,"Fichier Entrant")
        self.bg(self.convert,5,160,"Fichier Sortant")
        
        self.fonction_entrante()
        self.fonction_sortante()
        
        self.getsame_or_move = tkinter.StringVar()
        b=280
        liste_choix = {1:"Garder dans le meme dossier", 2:"Deplacer dans un autre"}
        for keys,values in liste_choix.items():
            self.choix__=tkinter.Radiobutton(self.convert,text=values, font=("arial",10), variable=self.getsame_or_move, value=keys)
            self.choix__.place(x=8,y=b)
            b+=20
            if keys == 2:
                self.choix__.config(command = self.GoTO_dir)
        self.getsame_or_move.set(1)

        self.format_compatible = ["png","jpeg","ico","jpg","bmp","dib","gif","jfif","jpe","tiff","tif"]
        
        self.bouton1__ = tkinter.Button(self.convert, text = "Ouvrir",relief=tkinter.GROOVE,width=10, font=("courier new",8), command=self.GoTO_dir)
        self.bouton1__.place(x=30, y=325)
        self.bouton2__ = tkinter.ttk.Button(self.convert, text = "Convertir",width=13, command=self.convertir)
        self.bouton2__.place(x=270, y=325)
        
    def From(self):
        self.From = askopenfile(filetypes=[("All","*.ico;*.png;*.jpg;*.bmp,*.dib;*.gif;*.jpeg;*.jfif;*.jpe;*.tiff;*.tif"),
                                        ("fichiers PNG","*.png"),
                                       ("Monochrome Bitmap","*.bmp;*.dib"),
                                       ("fichiers GIF","*.gif"),
                                       ("fichiers JPEG","*.jpg;*.jpeg;*.jfif;*.jpe)"),
                                       ("fichiers TIF","*.tif;*.tiff"),
                                        ("All","*.*"),
                                       ],title="Selectionne le fichier a convertir",initialdir=os.getcwd()[0:3])

    def GoTO_dir(self):
        self.getsame_or_move.set(2)
        self.GoTo = askdirectory(title="Choisis un emplacement",initialdir=os.path.dirname(os.getcwd()))
        
    def fonction_entrante(self):
        self.fichier_source = tkinter.StringVar()
        self.msg__ = tkinter.Checkbutton(self.convert, variable=self.fichier_source, text = "Conserver fichier source", font=("arial",12), bg="white")
        self.msg__.place(x=10, y=45)
        self.fichier_source.set(1)
        
        self.msg__ = tkinter.ttk.Button(self.convert, text = "Rechercher", width=13,command=self.From)
        self.msg__.place(x=270, y=90)

    def fonction_sortante(self):
        self.get_extension = tkinter.StringVar()
        x = 12
        y = 180
        for liste in ["png", "gif", "jpg","","tiff","bmp","ico"]:
            self.aff__=tkinter.Radiobutton(self.convert,text=liste.upper(), variable=self.get_extension, value=liste, bg ="white")
            if y <255:
                self.aff__.place(x=x,y=y)
                y+=25
            else:
                x = 112
                y = 180
        self.get_extension.set("png")

    def convertir(self):
        '''the convert, i can use **from PIL import Image** istead of cammand'''

        try:
            if self.From.name.split(".")[-1] not in self.format_compatible:
                showinfo("Erreur", "Fichier incompatible", icon='warning')
                self.From = ""
                self.Goto = ""
            else:
                rootname_ = self.From.name
                rootname = rootname_.replace("/","\\")
                nameOnly = os.path.basename(rootname)
                rootChemin = os.path.dirname(rootname)
                Name_without_Ext= ".".join(nameOnly.split(".")[:-1])
                if self.fichier_source.get() == "1" and self.getsame_or_move.get()=="1":
                    os.system("copy {0} {1}\\rescue.{2}".format(rootname,rootChemin,self.get_extension.get()))
                    os.system("rename {0}\\rescue.{1} {2}.{3}".format(rootChemin,self.get_extension.get(),\
                                                                    Name_without_Ext,self.get_extension.get()))
                    self.From = ""
                    self.Goto = ""
                    
                elif self.fichier_source.get() == "0" and self.getsame_or_move.get()== "1":
                    os.system("rename {0} {1}.{2}".format(rootname,Name_without_Ext,self.get_extension.get()))
                    self.From = ""
                    self.Goto = ""
                    
                elif self.fichier_source.get() == "1" and self.getsame_or_move.get()=="2":
                    if self.GoTo != "":
                        os.system("copy {0} {1}\\rescue.{2}".format(rootname,rootChemin,self.get_extension.get()))
                        os.system("rename {0}\\rescue.{1} {2}.{3}".format(rootChemin,self.get_extension.get(),\
                                                                        Name_without_Ext,self.get_extension.get()))
                        os.system("move {0}\\{1}.{2} {3}".format(rootChemin,Name_without_Ext,self.get_extension.get(),self.GoTo))
                        self.From = ""
                        self.Goto = ""
                    else:
                        showinfo("Alerte", "Veuillez fournir une destination\npour le fichier", icon='info')
                    
                elif self.fichier_source.get() == "0" and self.getsame_or_move.get()=="2":
                    if self.GoTo != "":
                        os.system("rename {0} {1}.{2}".format(rootname,Name_without_Ext,self.get_extension.get()))
                        os.system("move {0}\\{1}.{2} {3}".format(rootChemin,Name_without_Ext,self.get_extension.get(), self.GoTo))
                        self.From = ""
                        self.Goto = ""
                    else:
                        showinfo("Alerte", "Veuillez fournir une destination\npour le fichier", icon='info')
                else:
                    self.alerte(150,360)
                
        except AttributeError:
            showinfo("Alerte", "Vous devez selectionner un fichier source", icon='info')


isoFile.py
import tkinter
import os
import subprocess
from tkinter.filedialog import askopenfile,askdirectory,askopenfiles
from tkinter.messagebox import showinfo,askyesno,showerror,showwarning
import tkinter.ttk
import time
import threading
import string
import random

class IsoFile():
    def __init__(self,fen):
        ''' the main of part of the iso file converting
        Main de la parti de la conversion de fichier iso'''

        self.bout_loadimg = tkinter.PhotoImage(file="load.png") #Command : call the load_file function...
        self.bout_load = tkinter.ttk.Button(self.iso,image=self.bout_loadimg,command=self.load_file)# ^
        self.bout_load.place(x=300,y=40) 

        self.destin_img = tkinter.PhotoImage(file="dossier.png")
        self.destination = tkinter.ttk.Button(self.iso, image=self.destin_img,command=self.saveas)
        self.destination.place(x=310,y=260)

        self.type = tkinter.ttk.Combobox(self.iso, width=56, values=["UDF-ISO","ISO"],state="readonly")
        self.type.place(x=10,y=160)
        self.type.set("UDF-ISO")

        self.label = tkinter.Label(self.iso,text="")
        self.label.place(x=20,y=280)

        self.label = tkinter.Label(self.iso,text="Nom du label")
        self.label.place(x=25,y=260)
        
        self.get_Name=tkinter.StringVar()
        self.Name_label = tkinter.ttk.Entry(self.iso,width=15,textvariable=self.get_Name)
        self.Name_label.place(x=110,y=260)        

        self.choice_iso = tkinter.StringVar()
        liste2 = {2:"ISO Compression",1:"ISO Boot"}
        a = 220
        for key,texto in liste2.items():
            self.ISOboot_or_ISOcomp = tkinter.ttk.Radiobutton(self.iso, variable = self.choice_iso,text = texto, value=key)
            self.ISOboot_or_ISOcomp.place(x=a,y=50)
            a-= 120
        self.choice_iso.set("2")

        self.start_button = tkinter.Button(self.iso,borderwidth=2,bg="#%02x%02x%02x" % (71,86,82),fg="white", text = "Demarrer", command=lambda:self.conversion(fen))
        self.start_button.place(x=120,y=342)
        
        self.close_button = tkinter.Button(self.iso,borderwidth=2,bg="#%02x%02x%02x" % (71,86,82),fg="white", text = "Annuler", command=fen.destroy)
        self.close_button.place(x=200,y=342)
        
        liste = ["Inclut les fichiers et répertoires cachés.", "Utilise l’heure GMT pour tous les fichiers au lieu de l’heure locale.",""]
        self.choixHidden = tkinter.IntVar()
        self.choixGTL = tkinter.IntVar()
        
        self.check1=self.checkBox(5,185,self.choixHidden,liste[0]) #call the defined checkbox function
        self.check2=self.checkBox(5,205,self.choixGTL,liste[1]) #Appel a la fonction de checkbox definie

        self.choixOptimize = tkinter.IntVar()
        self.check3 = tkinter.Checkbutton(self.iso, text=liste[2],variable=self.choixOptimize,command=self.disable_orNot_3th_checkBox)
        self.check3.place(x=5,y=230) 

        self.image_clean = tkinter.PhotoImage(file="clean.png")
        self.label_clean = tkinter.ttk.Button(self.iso, image=self.image_clean,command=self.cleaning)
        self.label_clean.place(x=210,y=258)
        
        self.Logo()
        self.label_infos()
        
        self.get_optimize = tkinter.StringVar()
        self.Optimize(self.get_optimize) #Call the optimize choice function

        self.savefiles = "" #Pour permettre de creer la valeur, pour eviter l'erreur Attribute
        self.files = "" #Avoid the attribute error, when creating file, if it's empty (empty fields)
        
    def Logo(self):
        self.photo = tkinter.PhotoImage(file="iso2.png")
        self.logo = tkinter.Label(self.iso, image=self.photo)
        self.logo.place(x=10,y=10)

    def cleaning(self):
        '''When pressing the image_clean button:==> Clean the field'''
        self.get_Name.set("")
               
    def label_infos(self):
        '''generer des informations sur le fichier a compresser
        some informations about the file uploaded'''

        self.tab_infos = tkinter.LabelFrame(self.iso,text="Infos sur fichier",height=60,width=360,relief=tkinter.SUNKEN)
        self.tab_infos.place(x=10,y=90)
        self.info = tkinter.Label(self.iso,text="",fg="#%02x%02x%02x" % (96,00,00))
        self.info.place(x=20,y=105)
        self.info2 = tkinter.Label(self.iso,text="",fg="#%02x%02x%02x" % (96,00,00))
        self.info2.place(x=20,y=125)

    def Optimize(self,get_optimize):
        '''Optimize choices function'''

        self.optimize = tkinter.ttk.Combobox(self.iso, width=53, state=tkinter.DISABLED,values=["Optimiser avec l'algorithme de hachage MD5",\
                                                                         "Optimiser avec une comparaison binaire",\
                                                                         "Optimiser en codant les fichiers dupliqués une seule fois",],\
                                                                         textvariable=get_optimize)
        self.optimize.place(x=28,y=230)
        self.optimize.set("Optimiser avec l'algorithme de hachage MD5")
        
    def checkBox(self,x,y,get_choice,texte):
        '''Check box function, some properties'''

        self.choix = tkinter.Checkbutton(self.iso, text=texte,variable=get_choice)
        self.choix.place(x=x,y=y)
        
    def disable_orNot_3th_checkBox(self):
        '''function to disable the third choice within checkbox
        fonction qui est la pour disable ou enable le choix 3 dans les checkbox'''

        if self.choixOptimize.get() == 1:
            self.optimize.config(state=tkinter.NORMAL)
            self.optimize.config(state="readonly")
            
        else:
            self.optimize.config(state=tkinter.DISABLED)

    def lookingfor(self):
        '''On cherche si le fichier a compresser possede les fichiers propres pour une cle bootable
        looking for etfsboot.com file'''

        for dossier in os.listdir(self.files):
            if dossier == "boot":
                boo = self.files+r"boot/etfsboot.com"
                break
            else:
                boo = ""
        return boo
                
    def load_file(self):
        '''Ask the file to compress'''
        self.files=askdirectory(title="Selectionne le dossier",initialdir=os.getcwd()[0:3])
        self.A = self.files
        if self.files != "": #if is empty 
            #Thread for avoiding to wait when the loading finished
            size = threading.Thread(target=self.infos_file)
            size.start()
            
    def infos_file(self):
        '''some design, instead of to print size file with bytes, print it with (kb,mb,Gb)
        how many files, directories, ;-) '''

        self.size_total = 0
        self.nbre_fichier = 0
        self.nbre_dossier = -1
        if os.path.isdir(self.files):
            for dirpath,dirnames,filenames in os.walk(self.files):
                self.nbre_dossier += 1
                for f in filenames:
                    if self.size_total < 16106127360:
                        fp = os.path.join(dirpath,f)
                        self.size_total += os.path.getsize(fp)
                        self.nbre_fichier += 1
                    else:
                        self.size_total = -1
                        break
                if self.size_total == -1:
                    break

        total = self.size_total

        if total > 0:
            fois = 0
            while total > 1024:
                fois += 1
                total /= 1024

            if fois == 0:
                self.size = str(total)+" Bytes"
            else:
                if fois == 1:
                    ext = "Kb"
                elif fois == 2:
                    ext = "Mb"
                else:
                    ext = "Gb"
                size1 = str(total).split(".")
                self.size = size1[0]+"."+size1[1][0]+" "+ext
        else:
            self.size = "Trop volumineux"
            
        try:
            self.out_info.destroy()
            self.out_file_size.destroy()
        except:
            pass
            
        self.info.config(text="Taille du fichier: "+str(self.size))
        self.info2.config(text=str(self.nbre_fichier)+" fichiers au total")

    def configuration(self):
        '''Return the list of configurations, hidden files or not, system file or not ...'''

        if self.choixHidden.get() == 1:
            h = "-h"
        else:
            h = ""
            
        if self.choixGTL.get() == 1:
            g = "-g"
        else:
            g = ""
            
        if self.type.get() == "UDF-ISO":
            typer = "-u1"
        else:
            typer = "-u2"
            
        if self.get_Name.get() != "":
            name = "-l"+self.get_Name.get().upper().replace(" ","")
        else:
            name = ""
            
        if self.choixOptimize.get() == 1:
            if self.get_optimize.get() == "Optimiser avec l'algorithme de hachage MD5":
                opt = "-o"
            elif self.get_optimize.get() == "Optimiser avec une comparaison binaire":
                opt = "-oc"
            else:
                opt = "oi"
        else:
            opt = ""
        config = ["oscdimg.exe","-m"]
        li = [h,name,g,typer,opt]
        for args in li:
            if args != "":
                config.append(args)
                
        return config #Return la liste des configurations 

    def conversion(self,fen):
        '''Convert the directory to Iso '''

        try:
            if self.size != "Trop volumineux": # 
                if self.files != "":
                    if self.savefiles != "":
                        loop = self.lookingfor()
                        annuler = "no"
                        if self.choice_iso.get() == "1":
                            if loop == "": #Decide to create iso, we find etfsboot, we ask if we want to use this file
                                reponse = askyesno("Erreur etfsboot", "Vous avez choisi de faire un iso systeme bootable,qui est indisponible dans le fichier\n\
                                                   Voulez-vous le compresser tout de meme?")
                                if reponse == False:
                                    annuler = "yes"
                                else:
                                    loop = ""
                                                        
                        else: #if not
                            if loop != "":
                                reponse = askyesno("etfsboot trouve", "Vous avez choisi de compresser\n\
                                                   Voulez-vous le compresser en un systeme bootable?")
                                if reponse == False:
                                    loop = ""
                                else:
                                    pass

                        if annuler == "no": #if cancel == no
                            #The progess bar window
                            progress = tkinter.Toplevel()
                            progress.geometry("380x70+{0}+{1}".format(str(fen.winfo_x()),str(int(fen.winfo_y()+220))))
                            progress.title("Conversion...")
                            progress.minsize(380,70)
                            progress.maxsize(380,70)
                            progress.iconbitmap("bmp.ico")
                            progress.grab_set()

                            self.progress = tkinter.ttk.Progressbar(progress,orient="horizontal",length=361, mode="determinate")
                            self.progress.place(x=10,y=35)

                            self.cancel = tkinter.ttk.Button(progress, image=self.image_clean,command=progress.destroy)
                            self.cancel.place(x=340,y=5)

                            cmd = self.configuration()
                            
                            self.label_purcent = tkinter.Label(progress, text="")
                            self.label_purcent.place(x=175,y=10)
                            start = 0
                            niveau = 0
                            if loop != "":
                                cmd.append(loop)
                            cmd.append(self.files)
                            self.name_control = self.code_name()
                            Name = self.get_Name.get()+self.name_control
                            if Name == self.name_control:
                                Name = os.getlogin()+self.name_control
                            full_name = os.path.basename(self.files)+"_"+Name+".iso"
                            #if self.if_iso_already_exist(full_name,os.path.dirname(self.files)):
                                #full_name = os.path.basename(self.files)+"_"+Name+"@.iso"
                            cmd.append(self.savefiles+"/"+full_name)
                            dude = threading.Thread(target = self.ProgressBar, args=(niveau,cmd,Name,progress))
                            dude.start()

                    else:# that's mean the directory is not define, it's empty
                        question = askyesno('Emplacement non defini',"Voulez-vous enregistrer l'ISO au meme endroit ?")
                        if question == False:
                            self.saveas()
                        else: # wants to create the iso file to the same directory
                            self.savefiles = os.path.dirname(self.files)
                            self.conversion(fen)
                
                else: # Didn't choice folder to compress
                    showwarning("ERREUR","Dossier non defini")
            
            else: # Folder is too big , max 15 Gb
                showwarning("ERREUR","Le dossier est trop volumineux")
            
        except AttributeError as Error:
            showerror("ERREUR","Verifiez si c'est bien fait")
        except OSError:#Something wrong is happened   
            showinfo("ERREUR","Quelque chose de mal s'est passe !!!")             
                            
    def if_iso_already_exist(self,iso_name,directory):
        for file in os.listdir(directory):
            if file == iso_name:
                good = True
                break
            else:
                good = False
        return good

    def code_name(self):
        #Avoid some confusions with the name of generated iso file e.g : Iso_chfD54, Iso_Co0fd1
        letters = string.ascii_letters+string.digits
        list_of_code = [random.choice(letters) for _ in range(6)]
        return ''.join(list_of_code)

    def saveas(self): 
        '''get the directory for creating iso file'''
        if self.size != "Trop volumineux":
            self.savefiles=askdirectory(title="Enregistrer le fichier ISO",initialdir=os.getcwd()[0:3])
            self.B = self.savefiles
        else:
            showwarning("ERREUR","Le dossier est trop volumineux")
                
    def algo_tetChaje(self,line,niveau,pourcentage):
        '''Un peu complique, j'ai triche pour regler les pourcentages de progression,
        je ne sais pas comment bien faire simplement'''

        for i in range(1,101):
            if str(i)+"%" in line.rstrip().decode() and i not in pourcentage: #Get all lines from the cmd Dos
                pourcentage.append(i) # get the purcent from the command dos
                niveau = i
                break
        return niveau

    def out_file_info(self,Name,fenetre):
        '''Infos about the new iso file, the new size, the new name if didn't put the name_label'''
        try:
            fenetre.destroy()
            path_ = self.savefiles+"/"+os.path.basename(self.files)+"_"+Name+".iso"
            self.out_info= tkinter.Label(self.iso, text = "Chemin ---> "+path_,fg="blue")
            self.out_info.place(x=20,y=105)
            
            self.out_file_size = tkinter.Label(self.iso,text="Size de l'ISO ---> "+self.calcul_file(os.path.getsize(path_)),fg="blue")
            self.out_file_size.place(x=20,y=125)
            self.backNormal() #Call the back to normal function, clean all widgets
        except OSError as OE:
            showerror("ERREUR",OE) #Error
            
    def backNormal(self):
        '''Back to normal function ! The program can convert another folder'''

        self.info.config(text="")
        self.info2.config(text="")
        self.get_Name.set("")
        self.choixHidden.set(0)
        self.choixGTL.set(0)
        self.optimize.set("Optimiser avec l'algorithme de hachage MD5")
        self.choixOptimize.set(0)
        self.optimize.config(state=tkinter.DISABLED)
        self.type.set("UDF-ISO")
        self.savefiles = ""
        self.files = ""

    def calcul_file(self,iso):
        '''design the size of iso file, (kb,mb,gb)'''

        time = 0
        while iso > 1024:
            time += 1
            iso /= 1024

        if time == 0:
            retour = str(iso)+" Bytes"
        else:
            if time == 1:
                extension = "Kb"
            elif time == 2:
                extension = "Mb"
            else:
                extension = "Gb"
            retour1 = str(iso).split(".")
            retour = retour1[0]+"."+retour1[1][0]+" "+extension

        return retour

    def ProgressBar(self,niveau,cmd,Name,progress):
        '''Progress bar function, some text come from command dos, like the progress purcent'''

        try:
            g = subprocess.Popen(cmd,stdout=subprocess.PIPE, shell=True,stderr=subprocess.STDOUT, bufsize=1)
            pourcentage = []
            while True:
                self.progress["value"] = niveau
                line = g.stdout.readline(12)
                for i in range(1,101):
                    if str(i)+"%" in line.rstrip().decode() and i not in pourcentage:
                        pourcentage.append(i)
                        niveau = i
                        break
                if not line:
                    break
                self.progress.update()
                self.label_purcent.config(text=str(niveau)+"%", fg="#%02x%02x%02x" % (96,00,00))
        
            self.label_purcent.place(x=150,y=10)
            self.label_purcent.config(text="Complete",fg="blue")
            g.stdout.close()
            g.wait()
        
            affSorti=threading.Thread(target=self.out_file_info,args=(Name,progress))
            affSorti.start()

        except: # Cancel the iso compressing progress
            self.backNormal()
            showwarning("ATTENTION","Annuler une compresion ISO n'est pas bon pour votre ordinateur")


bootable.py
import tkinter,os
from tkinter.filedialog import askopenfile,askdirectory
from tkinter.messagebox import showinfo
import tkinter.ttk

class Bootable():
    ''''This part of code is not ready yet'''
    def __init__(self):
        self.bg(self.bootable,5,20,texte="Iso file",height=50)

        self.get_choice = tkinter.StringVar()
        self.liste = tkinter.ttk.Combobox(self.bootable,textvariable = self.get_choice,value=("NTFS","Fat32","FAT"),width=50)
        self.liste.place(x=25,y=154)

        self.get_choice1 = tkinter.StringVar()
        self.liste1 = tkinter.ttk.Combobox(self.bootable,textvariable = self.get_choice1,value=(""),width=50)
        self.liste1.place(x=25,y=204)

        self.get_choice2 = tkinter.StringVar()
        self.liste2 = tkinter.ttk.Combobox(self.bootable,textvariable = self.get_choice2,value=("NTFS","Fat32","FAT"),width=50)
        self.liste2.place(x=25,y=254)

        self.get_choice3 = tkinter.StringVar()
        self.liste3 = tkinter.ttk.Combobox(self.bootable,textvariable = self.get_choice3,value=("NTFS","Fat32","FAT"),width=50)
        self.liste3.place(x=25,y=304)

        


shutd.py
import tkinter
import os
import time
from subprocess import Popen,PIPE,STDOUT

class Shutd():
    def __init__(self,fen):
        self.bg(self.shutdown,5,160,"Dans combien de temps ?")
        self.bg(self.shutdown,5,20,"Eteindre ou Redemarrer ?")
        
        self.choice_get = tkinter.StringVar()
        liste = {"Shutdown":40,"Restart":230}
        for txt,key in liste.items():
            self.radio = tkinter.Radiobutton(self.shutdown,bg='white', text = txt, font=("arial",13),variable = self.choice_get, value=txt, anchor="w",pady=15)
            self.radio.place(x=key,y=50)
        
        self.message_alert = tkinter.Label(self.shutdown,text="", fg="red", font=("arial",14))
        self.message_alert.place(x=130,y=290)

        self.hour = tkinter.IntVar()
        self.time("Heure",20,210,60,210,self.hour)

        self.minute = tkinter.IntVar()
        self.time("Minute",134,210,180,210,self.minute)

        self.second = tkinter.IntVar()
        self.time("Seconde",255,210,308,210,self.second)
        
        self.bouton = tkinter.Button(self.shutdown, text = "Valider", relief=tkinter.GROOVE,command=lambda:self.validate(fen), width = 20)
        self.bouton.place(x=170,y=330)

        self.image()
        
    def time(self,label,a,b,x,y,time_get):
        '''timer board'''

        self.label_time = tkinter.Label(self.shutdown, text = label, fg="#%02x%02x%02x" % (71,86,82))
        self.label_time.place(x=a,y=b)
        self.spin = tkinter.Spinbox(self.shutdown, from_=0, to=60, textvariable = time_get, font=("arial",10),width=5)
        self.spin.place(x=x,y=y)

    def get_all_second(self):
        try:
            self.all_second = self.second.get()+((self.minute.get()+(self.hour.get()*60))*60)
        except ValueError:
            self.all_second=0
        return self.all_second

    def text_screen(self):
        if self.hour.get() == 0:
            if self.minute.get() == 0:
                self.text=self.choice_get.get()+" dans moins de: "+str(self.second.get())+" seconde(s)"
            elif self.minute.get() != 0 and self.second.get() == 0:
                self.text=self.choice_get.get()+" dans moins de: "+str(self.minute.get())+" minute(s)"
            elif self.minute.get() != 0 and self.second.get() != 0:
                self.text=self.choice_get.get()+" dans moins de: "+str(self.minute.get())+" minute(s) et "+str(self.second.get())+" seconde(s)"
        else:
            if self.minute.get() == 0:
                if self.second.get() ==0:
                    self.text=self.choice_get.get()+" dans moins de: "+str(self.hour.get())+" heure(s)"
                else:
                    self.text=self.choice_get.get()+" dans moins de: "+str(self.hour.get())+" heure(s) et "+str(self.second.get())+" seconde(s)"
            else:
                if self.text_second.get() ==0:
                    self.text=self.choice_get.get()+" dans moins de: "+str(self.hour.get())+" heure(s) et "+str(self.minute.get())+" minute(s)"
                else:
                    self.text=self.choice_get.get()+" dans moins de: "+str(self.hour.get())+" heure(s) et "+str(self.minute.get())+" minute(s) et "+str(self.second.get())+" seconde(s)"       
    
        return self.text

    def validate(self,fen):
        try:
            if (self.hour.get() == 0 and self.minute.get() ==0 and self.second.get() == 0):
                self.message_alert.config(text="Vous devez preciser l'heure")
            elif not self.choice_get.get():
                 self.message_alert.config(text="Restart or Shtdown ??")
            else:
                self.message_alert.destroy()
                self.bouton.destroy()
                cmd = ["Shutdown", '/'+self.choice_get.get()[0],'/f','/c',"'"+self.text_screen()+"'","/t",str(self.get_all_second())]
                h = Popen(cmd,stdout=PIPE, shell=True,stderr=STDOUT)
                with open("history", "a") as history:
                    history.write(time.strftime("%A %d %B %Y %H:%M:%S")+"\n"+str(self.get_all_second())+"\n"+self.choice_get.get()+"\n\n")
        except ValueError:
            self.message_alert.config(text="Format Heure Invalide")
                
    def image(self):
        self.photo = tkinter.PhotoImage(file="logo_shut.png")
        self.image = tkinter.Label(self.shutdown, image=self.photo)
        self.image.place(x=10,y=285)
